bin/utils.ml

```diff
@@ -3,7 +3,31 @@ let file_contents file = In_channel.(with_open_bin file In_channel.input_all)
 let pretty_print_link_status (link, (status_code, status_string)) =
   Printf.printf "%s - %i %s\n" link status_code status_string
 
-let pretty_print_link_status_from_file from_string extract_links file =
-  file |> file_contents |> from_string |> extract_links
-  |> List.iter (fun link ->
-         (link, Olinkcheck.Link.status link) |> pretty_print_link_status)
+let filter_with_ext ext = List.filter (fun file -> Filename.extension file = ext)
+
+let rec files_with_ext ext file =
+  if Sys.file_exists file then (
+    try
+      if Sys.is_directory file then
+        file |> Sys.readdir |> Array.to_list
+        |> List.map (Filename.concat file)
+        |> List.partition Sys.is_directory
+        |> fun (dirs, non_dirs) ->
+        filter_with_ext ext non_dirs @ List.concat_map (files_with_ext ext) dirs
+      else if Filename.extension file = ext then [ file ]
+      else []
+    with Sys_error x ->
+      print_endline (x ^ ", skipping.");
+      [])
+  else []
```

Comment:
Can we remove `if Sys.file_exists`? If false, we'll raise an error which is going to be matched. 

```suggestion
    try
      if Sys.is_directory file then
        file |> Sys.readdir |> Array.to_list
        |> List.map (Filename.concat file)
        |> List.partition Sys.is_directory
        |> fun (dirs, non_dirs) ->
        filter_with_ext ext non_dirs @ List.concat_map (files_with_ext ext) dirs
      else if Filename.extension file = ext then [ file ]
      else []
    with Sys_error x ->
      print_endline (x ^ ", skipping.");
      []
```

BTW, since you're writing an API, you should have `print_endline`

Reply:
> Can we remove `if Sys.file_exists`? If false, we'll raise an error which is going to be matched.
>

The problem isn't with the file given by the user itself - `Cmdliner` checks if that file exists. The problem is when there's some broken symlink somewhere. The `file_exists` check can be removed - `is_directory` will still raise an error in this case. Should we skip this and keep continuing with the other files or stop here by raising an error?

Reply:
Sounds good to me to skip bogus files or symlinks. Isn't it what grep -r does ?